Adding test to perform volume operations storm

divyenpatel opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:
Adding new test to perform volume operations storm
Test Steps
1. Create storage class for thin Provisioning.
2. Create 30 PVCs using above storage class in annotation, requesting 2 GB files.
3. Wait until all disks are ready and all PVs and PVCs get bind. (**CreateVolume** storm)
4. Create pod to mount volumes using PVCs created in step 2. (**AttachDisk** storm)
5. Wait for pod status to be running.
6. Verify all volumes accessible and available in the pod.
7. Delete pod.
8. wait until volumes gets detached. (**DetachDisk** storm)
9. Delete all PVCs. This should delete all Disks. (**DeleteVolume** storm)
10. Delete storage class.

This test will help validate issue reported at https://github.com/vmware/kubernetes/issues/71

**Which issue this PR fixes**
fixes #

**Special notes for your reviewer**:
executed test on 1.5.3 release with `VOLUME_OPS_SCALE` set to `5`
Will execute test with the changes made on PR - https://github.com/kubernetes/kubernetes/pull/42422, with the  `VOLUME_OPS_SCALE` set to `30`

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
None
```

cc: @abrarshivani @BaluDontu @tusharnt @pdhamdhere @luomiao @kerneltime

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42667)
<!-- Reviewable:end -->

divyenpatel commented:
@jeffvance Can you approve and review this PR?

liggitt commented:
how disruptive/expensive is this test to run? I don't have a good sense for what level of stress testing the volume e2es do by default

liggitt commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+	Steps
+    	1. Create storage class for thin Provisioning.
+    	2. Create 30 PVCs using above storage class in annotation, requesting 2 GB files.
+    	3. Wait until all disks are ready and all PVs and PVCs get bind. (CreateVolume strom)
```
s/strom/storm/g

jeffvance commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+	10. Delete storage class.
+*/
+
+var _ = framework.KubeDescribe("volume operations strom [Volume]", func() {
```
You probably wan the string "vsphere" in here.

jeffvance commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+			volume_ops_scale, err = strconv.Atoi(os.Getenv("VOLUME_OPS_SCALE"))
+			Expect(err).NotTo(HaveOccurred())
+		} else {
+			volume_ops_scale = 30
```
Nit: 30 could be a const

jeffvance commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("should create pod with many volumes and verify none attach call fails", func() {
```
change "none attach call" to "no attach calls"

jeffvance commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+		scParameters["diskformat"] = "thin"
+		storageClassSpec := getVSphereStorageClassSpec("thinsc", scParameters)
+		storageclass, err = client.StorageV1beta1().StorageClasses().Create(storageClassSpec)
+		Expect(err).NotTo(HaveOccurred())
```
would be nice to have creation of resources in separate funcs. And maybe moved to common/util code.

jeffvance commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+		for count < volume_ops_scale {
+			pvclaimSpec := getVSphereClaimSpecWithStorageClassAnnotation(namespace, storageclass)
+			claim, err := client.CoreV1().PersistentVolumeClaims(namespace).Create(pvclaimSpec)
+			Expect(err).NotTo(HaveOccurred())
```
why not use `framework.createPV`

jeffvance commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+			pvclaimSpec := getVSphereClaimSpecWithStorageClassAnnotation(namespace, storageclass)
+			claim, err := client.CoreV1().PersistentVolumeClaims(namespace).Create(pvclaimSpec)
+			Expect(err).NotTo(HaveOccurred())
+			pvclaims = append(pvclaims, claim)
```
nit: `pvclaims` could be pre-built with a capacity of `volume_ops_scale`. This would reduce go GC.

divyenpatel replied:
pre-allocated using 
```
		pvclaims = make([]*v1.PersistentVolumeClaim, volume_ops_scale)
		persistentvolumes = make([]*v1.PersistentVolume, volume_ops_scale)
```

jeffvance commented on test/e2e/BUILD:
```diff
@@ -101,6 +101,7 @@ go_library(
         "volumes.go",
         "vsphere_utils.go",
         "vsphere_volume_diskformat.go",
+        "vsphere_volume_ops_strom.go",
```
"strom" typo in filename

jeffvance commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+		By("Creating PVCs using the Storage Class")
+		count := 0
+		for count < volume_ops_scale {
+			pvclaimSpec := getVSphereClaimSpecWithStorageClassAnnotation(namespace, storageclass)
```
The storage class annotation is being deprecated and replaced by `Spec.storageClassName` in PVs, and claims.

divyenpatel replied:
tried using Spec.storageClassName on 1.5.3, but pvc is created without storage class name. since this is still being deprecated, for now using StorageClassAnnotation

jeffvance replied:
ok, probably will be available in the next release.

jeffvance commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+
+		By("Waiting for all claims to be in bound phase")
+		for _, claim := range pvclaims {
+			err = framework.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, client, namespace, claim.Name, framework.Poll, framework.ClaimProvisionTimeout)
```
`framework.WaitOnPVandPVC` performs similar. But you need to pass it a pv and a claim pair.

jeffvance replied:
can most of the body of the `for` loop be moved into a helper func?

divyenpatel replied:
moved to helper func

`func verifyVolumesAccessible(pod *v1.Pod, persistentvolumes []*v1.PersistentVolume, vsp *vsphere.VSphere)`

`func waitForPVClaimBoundPhase(client clientset.Interface, pvclaims []*v1.PersistentVolumeClaim) []*v1.PersistentVolume`

jeffvance commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+		podSpec.Spec.Volumes = volumes
+
+		pod, err := client.CoreV1().Pods(namespace).Create(podSpec)
+		Expect(err).NotTo(HaveOccurred())
```
again, above pod create code should be its own func

jeffvance commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Waiting for pod to be running")
+		Expect(framework.WaitForPodNameRunningInNamespace(client, pod.Name, namespace)).To(Succeed())
```
again, above pod create and wait code should be its own func, The idea is to make the It block short and readable.

jeffvance commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+
+		// get fresh pod info
+		pod, err = client.CoreV1().Pods(namespace).Get(pod.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
```
The above pod create/wait/Get code should be its own func, The idea is to make the It block short and readable.

divyenpatel replied:
moved to func
`func createPod(client clientset.Interface, namespace string, pvclaims []*v1.PersistentVolumeClaim) *v1.Pod`

jeffvance commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+			Expect(err).NotTo(HaveOccurred())
+			Expect(isAttached).To(BeTrue(), fmt.Sprintf("disk %v is not attached with the node", pv.Spec.VsphereVolume.VolumePath))
+			// Verify Volumes are accessible
+			filePath := "/mnt/" + fmt.Sprintf("volume%v", index+1) + "/emptyFile.txt"
```
preferred to use go's `Join` func for constructing path names.

divyenpatel replied:
used 
`filepath := filepath.Join("/mnt/", fmt.Sprintf("volume%v", index+1), "/emptyFile.txt")`

jeffvance commented on test/e2e/vsphere_volume_ops_strom.go:
```diff
@@ -0,0 +1,188 @@
+		}
+
+		By("Deleting pod")
+		err = client.CoreV1().Pods(namespace).Delete(pod.Name, nil)
```
can use `framework.DeletePodWithWait` to be sure that the pod is really deleted.

divyenpatel committed:
Adding test to perform volume operations storm

divyenpatel commented:
@liggitt @jeffvance addressed review comments.
@liggitt this is expensive test, but we will be running it against vsphere cloud provider only. 

We have some logic in vsphere cloud provider to select disk adapter based on available units, and SCSIControllerDeviceLimit. - https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/vsphere/vsphere.go#L60

so need some e2e test to cover this code.

jeffvance commented on test/e2e/vsphere_volume_ops_storm.go:
```diff
@@ -0,0 +1,205 @@
+
+// wait until all pvcs phase set to bound
+func waitForPVClaimBoundPhase(client clientset.Interface, pvclaims []*v1.PersistentVolumeClaim) []*v1.PersistentVolume {
+	var persistentvolumes []*v1.PersistentVolume
```
can simplify as: 
`   persistentvolumes := make([]*v1.PersistentVolume, len(pvclaims))`

divyenpatel replied:
done.

jeffvance commented:
@k8s-bot test this

jeffvance commented:
@k8s-bot gce etcd3 e2e test this

jeffvance commented:
@k8s-bot test this

jeffvance commented:
/lgtm

jeffvance commented:
/assign @jeffvance

jeffvance commented:
/approve

deads2k commented:
/approve

divyenpatel commented:
@jeffvance Can you add back lgtm? looks like rebase removed that label.

jeffvance commented on test/e2e/vsphere_volume_ops_storm.go:
```diff
@@ -0,0 +1,203 @@
+})
+
+// create pod with given claims
+func createPod(client clientset.Interface, namespace string, pvclaims []*v1.PersistentVolumeClaim) *v1.Pod {
```
This func is more flexible than the existing `CreateClientPod` and `MakePod` in _framework/_pv_util.go_. I know you want this pr merged very soon, but what do you think about moving this func to _pv_util_ and fixing up the calls in _persistent_volumes.go_ and _persistent_volumes-vsphere.go_?

divyenpatel replied:
Moved logic for creating pod spec with multiple PVCs to `pv_util -> MakePod`
Also updated `pv_util  -> CreateClientPod `to use newly added `pv_util -> CreatePod`

jeffvance commented on test/e2e/vsphere_volume_ops_storm.go:
```diff
@@ -0,0 +1,203 @@
+}
+
+// verify volumes are attached to the node and are accessible in pod
+func verifyVolumesAccessible(pod *v1.Pod, persistentvolumes []*v1.PersistentVolume, vsp *vsphere.VSphere) {
```
This func is in the e2e package and available to all code there. Maybe it can be moved to common/ or framework/, or maybe "vsphere" can be added to the func name?

divyenpatel replied:
Moved it to `vsphere_utils -> verifyVSphereVolumesAccessible`

jeffvance commented on test/e2e/vsphere_volume_ops_storm.go:
```diff
@@ -0,0 +1,203 @@
+}
+
+// wait until all pvcs phase set to bound
+func waitForPVClaimBoundPhase(client clientset.Interface, pvclaims []*v1.PersistentVolumeClaim) []*v1.PersistentVolume {
```
another useful helper function, especially for dynamic provisioning where we don't know the pv names. There is nothing specific to vsphere here so perhaps this can be moved to _framework/pv_util.go_?

divyenpatel replied:
Moved this func to `pv_util -> WaitForPVClaimBoundPhase`

divyenpatel committed:
addressed jeffvance's review comments

jeffvance commented:
/lgtm
